aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -61,6 +76,39 @@ static inline Tensor to_impl(const Tensor& self, const TensorOptions& options, b
   return r;
 }
 
+template <typename T>
+static inline bool is_null_or_equal_to(const c10::optional<T>& test, const T& value) {
+  if (!test.has_value()) {
+    return true;
+  }
+  return test.value() == value;
+}
+
+static inline Tensor to_impl(
+    const Tensor& self,
+    c10::optional<ScalarType> dtype,
+    c10::optional<Layout> layout,
+    c10::optional<Device> device,
+    c10::optional<bool> pin_memory,
+    bool non_blocking,
+    bool copy,
+    c10::optional<c10::MemoryFormat> optional_memory_format) {
+  auto memory_format = optional_memory_format.value_or(MemoryFormat::Preserve);
+
+  // fast path
+  if (is_null_or_equal_to(dtype, self.dtype().toScalarType()) &&
+      is_null_or_equal_to(layout, self.layout()) &&
+      is_null_or_equal_to(device, self.device()) &&
+      !copy &&
+      (memory_format == MemoryFormat::Preserve ||
```

Comment:
Why not `is_null_or_equal_to(optional_memory_format, MemoryFormat::Preserve)` ?
Also I'm not sure what is the suggest case below doing?

Reply:
> Why not is_null_or_equal_to(optional_memory_format, MemoryFormat::Preserve)

No reason, I can change it

> Also I'm not sure what is the suggest case below doing?

I copied this from the original `to_impl` code, but here's the explanation:
- The suggest case is saying that "if self has the same memory_format as the one passed to `to_impl`, then we don't need to do anything"
- Unfortunately we can't just query `self.options().memory_format()` and compare it to `optional_memory_format` because of a quirk of how MemoryFormat is implemented (see next two bullet points)
- Sometimes a ChannelsLast tensor doesn't actually get marked as being ChannelsLast.
- In order to tell if a ChannelsLast tensor is actually ChannelsLast, we have to do `self.suggest_memory_format(memory_format)` or `self.is_contiguous(memory_format=memory_format)` and this code is doing the former

Reply:
ok!
Thanks for the full explanation! Is that captured somewhere in a comment in code?

Reply:
I don't think so, I'll put a comment in about this over here

Reply:
Reading this code more... the suggest case seems like a bug: https://github.com/pytorch/pytorch/issues/62027 .
Since the code as-is reproduces the behavior before the refactor, is it fine if I leave it as-is for now and then fix it whenever #62027 gets a resolution?